File: skills/adkit/scripts/src/lib/cluster.test.ts
/**
 * Tests for the data-driven cluster analysis shared by report and audit.
 *
 * Pure-function tests — feed performance rows, assert the proposals. No IO.
 */

import { describe, expect, it } from "vitest";
import {
  clusterSplitRecommendation,
  keywordsByClicksAndCtr,
  keywordsToPromote,
  negativesToAdd,
  type Row,
} from "./cluster.js";

interface StOptions {
  clicks?: number;
  conversions?: number;
  cost?: number;
  impressions?: number;
  adGroupId?: string;
}

function st(term: string, options: StOptions = {}): Row {
  const { clicks = 0, conversions = 0, cost = 0, impressions = 0, adGroupId = "1" } = options;
  return {
    search_term: term,
    ad_group_id: adGroupId,
    clicks,
    conversions,
    cost,
    impressions,
  };
}

describe("keywordsToPromote", () => {
  it("promotes converting and clicked terms", () => {
    const rows = [
      st("online reputation software", { clicks: 5, conversions: 2, cost: 40 }),
      st("reputation monitoring", { clicks: 4, conversions: 0, cost: 12 }),
      st("free review widget", { clicks: 1, conversions: 0, cost: 1 }), // below bar
    ];
    const out = keywordsToPromote(rows);
    const texts = out.map((p) => p.text);
    expect(texts).toEqual(["online reputation software", "reputation monitoring"]);
    expect(out.every((p) => p.matchType === "PHRASE")).toBe(true);
  });

  it("excludes existing keywords case-insensitively", () => {
    const rows = [st("Online Reputation Software", { clicks: 9, conversions: 3, cost: 50 })];
    const out = keywordsToPromote(rows, [{ text: "online reputation software" }]);
    expect(out).toEqual([]);
  });

  it("aggregates same term across ad groups", () => {
    const rows = [
      st("review software", { clicks: 2, conversions: 0, cost: 6, adGroupId: "1" }),
      st("review software", { clicks: 2, conversions: 1, cost: 8, adGroupId: "2" }),
    ];
    const out = keywordsToPromote(rows);
    expect(out.length).toBe(1);
    expect(out[0].clicks).toBe(4);
    expect(out[0].conversions).toBe(1.0);
    expect(out[0].cost).toBe(14.0);
  });

  it("sorted strongest first and capped by limit", () => {
    const rows = Array.from({ length: 5 }, (_, i) =>
      st(`kw ${i}`, { clicks: 10, conversions: i }),
    );
    const out = keywordsToPromote(rows, [], { limit: 2 });
    expect(out.map((p) => p.text)).toEqual(["kw 4", "kw 3"]);
  });
});

describe("negativesToAdd", () => {
  it("flags zero-conversion spend", () => {
    const rows = [
      st("cheap reviews free", { clicks: 6, conversions: 0, cost: 9 }),
      st("reputation software pricing", { clicks: 3, conversions: 2, cost: 20 }), // converted -> keep
    ];
    const out = negativesToAdd(rows);
    expect(out.map((n) => n.text)).toEqual(["cheap reviews free"]);
    expect(out[0].cost).toBe(9.0);
  });

  it("ignores terms under min cost", () => {
    const rows = [st("barely spent", { clicks: 1, conversions: 0, cost: 0.4 })];
    expect(negativesToAdd(rows, { minCost: 1.0 })).toEqual([]);
  });

  it("sorted by wasted cost desc", () => {
    const rows = [
      st("waste a", { clicks: 2, conversions: 0, cost: 5 }),
      st("waste b", { clicks: 9, conversions: 0, cost: 30 }),
    ];
    const out = negativesToAdd(rows);
    expect(out.map((n) => n.text)).toEqual(["waste b", "waste a"]);
  });
});

describe("keywordsByClicksAndCtr", () => {
  it("ranks by clicks/CTR, ignoring conversions, and flags cheap CPC as strong", () => {
    const rows = [
      // high clicks, high CTR, cheap CPC -> strong + cheapToScale
      st("best running shoes", { clicks: 10, impressions: 100, cost: 5, conversions: 0 }),
      // high clicks, CTR far below the campaign average -> weak-relevance
      st("running shoes", { clicks: 5, impressions: 1000, cost: 5, conversions: 9 }),
      // high clicks, above-average CTR, but CPC is not cheap -> watch
      st("premium running shoes", { clicks: 8, impressions: 20, cost: 20 }),
    ];
    const out = keywordsByClicksAndCtr(rows, [], 1.0);
    const byText = Object.fromEntries(out.map((c) => [c.text, c]));
    expect(byText["best running shoes"].verdict).toBe("strong");
    expect(byText["best running shoes"].cheapToScale).toBe(true);
    expect(byText["running shoes"].verdict).toBe("weak-relevance");
    expect(byText["premium running shoes"].verdict).toBe("watch");
    expect(byText["premium running shoes"].cheapToScale).toBe(false);
    // sorted by clicks desc
    expect(out.map((c) => c.text)).toEqual([
      "best running shoes",
      "premium running shoes",
      "running shoes",
    ]);
  });

  it("treats campaignAvgCpc=0 as unknown and skips the CPC check", () => {
    const rows = [st("wide shoe fit", { clicks: 10, impressions: 20, cost: 500 })];
    const out = keywordsByClicksAndCtr(rows, [], 0);
    expect(out[0].verdict).toBe("strong");
    expect(out[0].cheapToScale).toBe(false);
  });

  it("keeps the same term separate per ad group instead of merging", () => {
    const rows = [
      st("shoe repair", { clicks: 4, impressions: 40, adGroupId: "111" }),
      st("shoe repair", { clicks: 6, impressions: 60, adGroupId: "222" }),
    ];
    const out = keywordsByClicksAndCtr(rows, []);
    expect(out.length).toBe(2);
    expect(out.map((c) => c.adGroupId).sort()).toEqual([111, 222]);
  });

  it("excludes existing keywords and terms under minClicks", () => {
    const rows = [
      st("already a keyword", { clicks: 20, impressions: 40 }),
      st("too few clicks", { clicks: 1, impressions: 2 }),
    ];
    const out = keywordsByClicksAndCtr(rows, [{ text: "Already A Keyword" }], 0, {
      minClicks: 3,
    });
    expect(out).toEqual([]);
  });
});

describe("clusterSplitRecommendation", () => {
  function kw(text: string, cpc: number): Row {
    return { text, avg_cpc: cpc };
  }

  it("recommends split on wide cpc spread", () => {
    const kws = [
      kw("client engagement software", 0.99),
      kw("engagement tool", 1.1),
      kw("online reputation software", 12.0),
      kw("reputation management software", 18.0),
    ];
    const rec = clusterSplitRecommendation(kws);
    expect(rec).not.toBeNull();
    expect(rec!.ratio).toBeGreaterThanOrEqual(3.0);
    expect(rec!.expensive).toContain("online reputation software");
    expect(rec!.cheap).toContain("client engagement software");
  });

  it("none when spread is tight", () => {
    const kws = Array.from({ length: 5 }, (_, i) => kw(`kw ${i}`, 2.0 + i * 0.1));
    expect(clusterSplitRecommendation(kws)).toBeNull();
  });

  it("none when too few priced keywords", () => {
    const kws = [kw("a", 1.0), kw("b", 10.0), { text: "c", avg_cpc: 0 }];
    expect(clusterSplitRecommendation(kws, { minKeywords: 4 })).toBeNull();
  });
});

File: skills/adkit/scripts/ads_skill/lib/cluster_test.py
"""Tests for the data-driven cluster analysis shared by report and audit.

Pure-function tests — feed performance rows, assert the proposals. No IO."""

from __future__ import annotations

from ads_skill.lib.cluster import (
    cluster_split_recommendation,
    keywords_to_promote,
    negatives_to_add,
)


def _st(term: str, *, clicks: int = 0, conversions: float = 0.0, cost: float = 0.0,
        impressions: int = 0, ad_group_id: str = "1") -> dict:
    return {
        "search_term": term, "ad_group_id": ad_group_id, "clicks": clicks,
        "conversions": conversions, "cost": cost, "impressions": impressions,
    }


class TestKeywordsToPromote:
    def test_promotes_converting_and_clicked_terms(self) -> None:
        rows = [
            _st("online reputation software", clicks=5, conversions=2, cost=40),
            _st("reputation monitoring", clicks=4, conversions=0, cost=12),
            _st("free review widget", clicks=1, conversions=0, cost=1),  # below bar
        ]
        out = keywords_to_promote(rows)
        texts = [p["text"] for p in out]
        assert texts == ["online reputation software", "reputation monitoring"]
        assert all(p["matchType"] == "PHRASE" for p in out)

    def test_excludes_existing_keywords_case_insensitively(self) -> None:
        rows = [_st("Online Reputation Software", clicks=9, conversions=3, cost=50)]
        out = keywords_to_promote(rows, [{"text": "online reputation software"}])
        assert out == []

    def test_aggregates_same_term_across_ad_groups(self) -> None:
        rows = [
            _st("review software", clicks=2, conversions=0, cost=6, ad_group_id="1"),
            _st("review software", clicks=2, conversions=1, cost=8, ad_group_id="2"),
        ]
        out = keywords_to_promote(rows)
        assert len(out) == 1
        assert out[0]["clicks"] == 4
        assert out[0]["conversions"] == 1.0
        assert out[0]["cost"] == 14.0

    def test_sorted_strongest_first_and_capped_by_limit(self) -> None:
        rows = [_st(f"kw {i}", clicks=10, conversions=i) for i in range(5)]
        out = keywords_to_promote(rows, limit=2)
        assert [p["text"] for p in out] == ["kw 4", "kw 3"]


class TestNegativesToAdd:
    def test_flags_zero_conversion_spend(self) -> None:
        rows = [
            _st("cheap reviews free", clicks=6, conversions=0, cost=9),
            _st("reputation software pricing", clicks=3, conversions=2, cost=20),  # converted → keep
        ]
        out = negatives_to_add(rows)
        assert [n["text"] for n in out] == ["cheap reviews free"]
        assert out[0]["cost"] == 9.0

    def test_ignores_terms_under_min_cost(self) -> None:
        rows = [_st("barely spent", clicks=1, conversions=0, cost=0.4)]
        assert negatives_to_add(rows, min_cost=1.0) == []

    def test_sorted_by_wasted_cost_desc(self) -> None:
        rows = [
            _st("waste a", clicks=2, conversions=0, cost=5),
            _st("waste b", clicks=9, conversions=0, cost=30),
        ]
        out = negatives_to_add(rows)
        assert [n["text"] for n in out] == ["waste b", "waste a"]


class TestClusterSplitRecommendation:
    def _kw(self, text: str, cpc: float) -> dict:
        return {"text": text, "avg_cpc": cpc}

    def test_recommends_split_on_wide_cpc_spread(self) -> None:
        kws = [
            self._kw("client engagement software", 0.99),
            self._kw("engagement tool", 1.10),
            self._kw("online reputation software", 12.00),
            self._kw("reputation management software", 18.00),
        ]
        rec = cluster_split_recommendation(kws)
        assert rec is not None
        assert rec["ratio"] >= 3.0
        assert "online reputation software" in rec["expensive"]
        assert "client engagement software" in rec["cheap"]

    def test_none_when_spread_is_tight(self) -> None:
        kws = [self._kw(f"kw {i}", 2.0 + i * 0.1) for i in range(5)]
        assert cluster_split_recommendation(kws) is None

    def test_none_when_too_few_priced_keywords(self) -> None:
        kws = [self._kw("a", 1.0), self._kw("b", 10.0), {"text": "c", "avg_cpc": 0}]
        assert cluster_split_recommendation(kws, min_keywords=4) is None
